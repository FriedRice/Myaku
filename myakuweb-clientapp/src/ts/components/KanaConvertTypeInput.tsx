/** @module Romaji->kana conversion setting input component */

import React from 'react';

import {
    KANA_CONVERT_TYPE_VALUES,
    KanaConvertType,
} from '../types';

interface KanaConvertTypeInputProps {
    kanaConvertType: KanaConvertType;
    onChange: (kanaConvertType: KanaConvertType) => void;
}
type Props = KanaConvertTypeInputProps;

const CONVERT_TYPE_LABELS: Record<KanaConvertType, string> = {
    'hira': 'Hiragana (a→あ)',
    'kata': 'Katakana (a→ア)',
    'none': 'No conversion (a→a)',
};


function createRadioInputs(
    selected: KanaConvertType,
    handleChangeFunc: (event: React.FormEvent<HTMLInputElement>) => void
): React.ReactElement[] {
    var radioInputs: React.ReactElement[] = [];
    for (const convertType of KANA_CONVERT_TYPE_VALUES) {
        const inputId = `kana-conv-${convertType}-radio`;
        radioInputs.push(
            <div key={inputId} className='search-options-check'>
                <input
                    className='search-options-check-input'
                    id={inputId}
                    type='radio'
                    name='kana-convert-type'
                    value={convertType}
                    checked={convertType === selected}
                    onChange={handleChangeFunc}
                />
                <label
                    className='check-input-label'
                    htmlFor={inputId}
                >
                    {CONVERT_TYPE_LABELS[convertType]}
                </label>
            </div>
        );
    }

    return radioInputs;
}

const KanaConvertTypeInput: React.FC<Props> = function(props) {
    function handleChange(event: React.FormEvent<HTMLInputElement>): void {
        props.onChange(event.currentTarget.value as KanaConvertType);
    }

    return (
        <fieldset className='kana-conv-field'>
            <legend className='search-options-legend'>
                Romaji Conversion
            </legend>
            {createRadioInputs(props.kanaConvertType, handleChange)}
        </fieldset>
    );
};

export default KanaConvertTypeInput;
