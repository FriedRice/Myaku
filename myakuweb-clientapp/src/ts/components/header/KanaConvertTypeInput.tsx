/** @module Romaji->kana conversion setting input component */

import React from 'react';
import useInputChangeHandler from 'ts/hooks/useInputChangeHandler';

import {
    KANA_CONVERT_TYPE_VALUES,
    KanaConvertType,
} from 'ts/types/types';

interface KanaConvertTypeInputProps {
    kanaConvertType: KanaConvertType;
    onChange: (kanaConvertType: KanaConvertType) => void;
}
type Props = KanaConvertTypeInputProps;

const CONVERT_TYPE_LABELS: Record<KanaConvertType, string> = {
    'hira': 'Hiragana (a→あ)',
    'kata': 'Katakana (a→ア)',
    'none': 'No conversion (a→a)',
};


function createRadioInputs(
    selected: KanaConvertType,
    handleChangeFunc: (event: React.FormEvent<HTMLInputElement>) => void
): React.ReactElement[] {
    var radioInputs: React.ReactElement[] = [];
    for (const convertType of KANA_CONVERT_TYPE_VALUES) {
        const inputId = `kana-conv-${convertType}-radio`;
        radioInputs.push(
            <div key={inputId} className='search-options-check'>
                <input
                    className='search-options-check-input'
                    id={inputId}
                    type='radio'
                    name='kana-convert-type'
                    value={convertType}
                    checked={convertType === selected}
                    onChange={handleChangeFunc}
                />
                <label
                    className='check-input-label'
                    htmlFor={inputId}
                >
                    {CONVERT_TYPE_LABELS[convertType]}
                </label>
            </div>
        );
    }

    return radioInputs;
}

const KanaConvertTypeInput: React.FC<Props> = function(props) {
    const handleChange = useInputChangeHandler(props.onChange);

    return (
        <fieldset className='kana-conv-field'>
            <legend className='search-options-legend'>
                Romaji Conversion
            </legend>
            {createRadioInputs(props.kanaConvertType, handleChange)}
        </fieldset>
    );
};

export default KanaConvertTypeInput;
