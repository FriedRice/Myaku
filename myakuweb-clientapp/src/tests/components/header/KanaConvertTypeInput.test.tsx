/**
 * Tests for the [[KanaConvertTypeInput]] component.
 */

import KanaConvertTypeInput from 'ts/components/header/KanaConvertTypeInput';
import React from 'react';

import {
    HTMLAttributes,
    ReactWrapper,
    mount,
} from 'enzyme';
import {
    KanaConvertType,
    isKanaConvertType,
} from 'ts/types/types';

const LABEL_TEXT_MAP: Record<KanaConvertType, string> = {
    'hira': 'Hiragana',
    'kata': 'Katakana',
    'none': 'No conversion',
};

function expectRadioChecked(
    wrapper: ReactWrapper, checkedValue: KanaConvertType
): void {
    var seenValues: string[] = [];
    var checkedValueFound = false;
    wrapper.find('input').forEach(
        function(node: ReactWrapper<HTMLAttributes>) {
            const nodeValue = node.props().value;
            if (typeof nodeValue !== 'string') {
                throw new Error(
                    `Radio input value is not a string: ${nodeValue}`
                );
            }
            expect(seenValues).not.toContain(nodeValue);
            seenValues.push(nodeValue);

            if (nodeValue === checkedValue) {
                checkedValueFound = true;
                expect(node.props().checked).toBe(true);
            } else {
                expect(node.props().checked).toBe(false);
            }
        }
    );

    expect(checkedValueFound).toBe(true);
}

function simulateRadioInputChange(
    wrapper: ReactWrapper, value: KanaConvertType
): void {
    const inputWrapper = wrapper.find(`input[value="${value}"]`);
    expect(inputWrapper).toHaveLength(1);

    inputWrapper.simulate('change');
}


describe('<KanaConvertTypeInput />', function() {
    it('has correct labels for inputs', function() {
        const wrapper = mount(
            <KanaConvertTypeInput
                kanaConvertType={'hira'}
                onChange={(): void => {}}
            />
        );
        wrapper.find('input').forEach(
            function(node: ReactWrapper<HTMLAttributes>) {
                const label = wrapper.find(
                    `label[htmlFor="${node.props().id}"]`
                );
                expect(label).toHaveLength(1);

                const nodeValue = node.props().value;
                if (!isKanaConvertType(nodeValue)) {
                    throw new Error(
                        'Radio input value is not a valid kana convert '
                        + `type: "${nodeValue}"`
                    );
                }
                expect(label.text()).toStrictEqual(
                    expect.stringContaining(LABEL_TEXT_MAP[nodeValue])
                );
            }
        );
    });

    it('renders correct checked input', function() {
        const wrapper = mount(
            <KanaConvertTypeInput
                kanaConvertType={'hira'}
                onChange={(): void => {}}
            />
        );
        expect(wrapper).toMatchSnapshot();
        expectRadioChecked(wrapper, 'hira');

        wrapper.setProps({kanaConvertType: 'kata'});
        expect(wrapper).toMatchSnapshot();
        expectRadioChecked(wrapper, 'kata');

        wrapper.setProps({kanaConvertType: 'none'});
        expect(wrapper).toMatchSnapshot();
        expectRadioChecked(wrapper, 'none');

        wrapper.setProps({kanaConvertType: 'hira'});
        expect(wrapper).toMatchSnapshot();
        expectRadioChecked(wrapper, 'hira');
    });

    it('calls onChange on radio input change', function() {
        const mockOnChange = jest.fn();
        const wrapper = mount(
            <KanaConvertTypeInput
                kanaConvertType={'hira'}
                onChange={mockOnChange}
            />
        );
        simulateRadioInputChange(wrapper, 'kata');
        expect(mockOnChange).toBeCalledTimes(1);
        expect(mockOnChange).lastCalledWith('kata');

        simulateRadioInputChange(wrapper, 'none');
        expect(mockOnChange).toBeCalledTimes(2);
        expect(mockOnChange).lastCalledWith('none');

        simulateRadioInputChange(wrapper, 'hira');
        expect(mockOnChange).toBeCalledTimes(3);
        expect(mockOnChange).lastCalledWith('hira');
    });
});
