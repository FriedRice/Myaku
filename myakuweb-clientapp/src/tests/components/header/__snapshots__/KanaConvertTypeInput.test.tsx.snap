// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KanaConvertTypeInput /> renders correct checked input 1`] = `
<KanaConvertTypeInput
  kanaConvertType="hira"
  onChange={[Function]}
>
  <fieldset
    className="kana-conv-field"
  >
    <legend
      className="search-options-legend"
    >
      Romaji Conversion
    </legend>
    <div
      className="search-options-check"
      key="kana-conv-hira-radio"
    >
      <input
        checked={true}
        className="search-options-check-input"
        id="kana-conv-hira-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="hira"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-hira-radio"
      >
        Hiragana (a→あ)
      </label>
    </div>
    <div
      className="search-options-check"
      key="kana-conv-kata-radio"
    >
      <input
        checked={false}
        className="search-options-check-input"
        id="kana-conv-kata-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="kata"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-kata-radio"
      >
        Katakana (a→ア)
      </label>
    </div>
    <div
      className="search-options-check"
      key="kana-conv-none-radio"
    >
      <input
        checked={false}
        className="search-options-check-input"
        id="kana-conv-none-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="none"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-none-radio"
      >
        No conversion (a→a)
      </label>
    </div>
  </fieldset>
</KanaConvertTypeInput>
`;

exports[`<KanaConvertTypeInput /> renders correct checked input 2`] = `
<KanaConvertTypeInput
  kanaConvertType="kata"
  onChange={[Function]}
>
  <fieldset
    className="kana-conv-field"
  >
    <legend
      className="search-options-legend"
    >
      Romaji Conversion
    </legend>
    <div
      className="search-options-check"
      key="kana-conv-hira-radio"
    >
      <input
        checked={false}
        className="search-options-check-input"
        id="kana-conv-hira-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="hira"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-hira-radio"
      >
        Hiragana (a→あ)
      </label>
    </div>
    <div
      className="search-options-check"
      key="kana-conv-kata-radio"
    >
      <input
        checked={true}
        className="search-options-check-input"
        id="kana-conv-kata-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="kata"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-kata-radio"
      >
        Katakana (a→ア)
      </label>
    </div>
    <div
      className="search-options-check"
      key="kana-conv-none-radio"
    >
      <input
        checked={false}
        className="search-options-check-input"
        id="kana-conv-none-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="none"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-none-radio"
      >
        No conversion (a→a)
      </label>
    </div>
  </fieldset>
</KanaConvertTypeInput>
`;

exports[`<KanaConvertTypeInput /> renders correct checked input 3`] = `
<KanaConvertTypeInput
  kanaConvertType="none"
  onChange={[Function]}
>
  <fieldset
    className="kana-conv-field"
  >
    <legend
      className="search-options-legend"
    >
      Romaji Conversion
    </legend>
    <div
      className="search-options-check"
      key="kana-conv-hira-radio"
    >
      <input
        checked={false}
        className="search-options-check-input"
        id="kana-conv-hira-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="hira"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-hira-radio"
      >
        Hiragana (a→あ)
      </label>
    </div>
    <div
      className="search-options-check"
      key="kana-conv-kata-radio"
    >
      <input
        checked={false}
        className="search-options-check-input"
        id="kana-conv-kata-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="kata"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-kata-radio"
      >
        Katakana (a→ア)
      </label>
    </div>
    <div
      className="search-options-check"
      key="kana-conv-none-radio"
    >
      <input
        checked={true}
        className="search-options-check-input"
        id="kana-conv-none-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="none"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-none-radio"
      >
        No conversion (a→a)
      </label>
    </div>
  </fieldset>
</KanaConvertTypeInput>
`;

exports[`<KanaConvertTypeInput /> renders correct checked input 4`] = `
<KanaConvertTypeInput
  kanaConvertType="hira"
  onChange={[Function]}
>
  <fieldset
    className="kana-conv-field"
  >
    <legend
      className="search-options-legend"
    >
      Romaji Conversion
    </legend>
    <div
      className="search-options-check"
      key="kana-conv-hira-radio"
    >
      <input
        checked={true}
        className="search-options-check-input"
        id="kana-conv-hira-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="hira"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-hira-radio"
      >
        Hiragana (a→あ)
      </label>
    </div>
    <div
      className="search-options-check"
      key="kana-conv-kata-radio"
    >
      <input
        checked={false}
        className="search-options-check-input"
        id="kana-conv-kata-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="kata"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-kata-radio"
      >
        Katakana (a→ア)
      </label>
    </div>
    <div
      className="search-options-check"
      key="kana-conv-none-radio"
    >
      <input
        checked={false}
        className="search-options-check-input"
        id="kana-conv-none-radio"
        name="kana-convert-type"
        onChange={[Function]}
        type="radio"
        value="none"
      />
      <label
        className="check-input-label"
        htmlFor="kana-conv-none-radio"
      >
        No conversion (a→a)
      </label>
    </div>
  </fieldset>
</KanaConvertTypeInput>
`;
